File: src/metrics/dora.ts
import { differenceInDays, differenceInHours } from 'date-fns';
import type {
  DoraMetrics,
  DeploymentFrequency,
  LeadTime,
  ChangeFailureRate,
  MTTR,
  DoraRating,
  MetricsPeriod,
  GitHubDeployment,
  GitHubPullRequest,
  LinearIssue,
  SlabPostmortem,
} from '../types/index.js';

export interface DoraCalculationInput {
  deployments: GitHubDeployment[];
  pullRequests: GitHubPullRequest[];
  incidents: LinearIssue[];
  postmortems: SlabPostmortem[];
  period: MetricsPeriod;
}

export function calculateDoraMetrics(input: DoraCalculationInput): DoraMetrics {
  const { deployments, pullRequests, incidents, postmortems, period } = input;

  return {
    deploymentFrequency: calculateDeploymentFrequency(deployments, period),
    leadTime: calculateLeadTime(pullRequests),
    changeFailureRate: calculateChangeFailureRate(deployments, incidents),
    mttr: calculateMTTR(incidents, postmortems),
    period,
    generatedAt: new Date(),
  };
}

function calculateDeploymentFrequency(
  deployments: GitHubDeployment[],
  period: MetricsPeriod
): DeploymentFrequency {
  const successfulDeployments = deployments.filter(d => d.status === 'success');
  const totalDeployments = successfulDeployments.length;

  const days = differenceInDays(period.endDate, period.startDate) || 1;
  const deploymentsPerDay = totalDeployments / days;
  const deploymentsPerWeek = deploymentsPerDay * 7;

  const rating = getDeploymentFrequencyRating(deploymentsPerDay);

  return {
    deploymentsPerDay,
    deploymentsPerWeek,
    totalDeployments,
    rating,
  };
}

function getDeploymentFrequencyRating(deploymentsPerDay: number): DoraRating {
  // Elite: Multiple deploys per day
  if (deploymentsPerDay >= 1) return 'elite';
  // High: Between once per day and once per week
  if (deploymentsPerDay >= 1 / 7) return 'high';
  // Medium: Between once per week and once per month
  if (deploymentsPerDay >= 1 / 30) return 'medium';
  // Low: Less than once per month
  return 'low';
}

function calculateLeadTime(pullRequests: GitHubPullRequest[]): LeadTime {
  const leadTimes: number[] = [];

  for (const pr of pullRequests) {
    if (!pr.mergedAt) continue;

    const startTime = pr.firstCommitAt ?? pr.createdAt;
    const hours = differenceInHours(pr.mergedAt, startTime);
    leadTimes.push(hours);
  }

  if (leadTimes.length === 0) {
    return {
      averageHours: 0,
      medianHours: 0,
      p90Hours: 0,
      rating: 'low',
    };
  }

  const sortedLeadTimes = [...leadTimes].sort((a, b) => a - b);
  const averageHours = leadTimes.reduce((a, b) => a + b, 0) / leadTimes.length;
  const medianHours = sortedLeadTimes[Math.floor(sortedLeadTimes.length / 2)];
  const p90Index = Math.floor(sortedLeadTimes.length * 0.9);
  const p90Hours = sortedLeadTimes[p90Index] ?? medianHours;

  const rating = getLeadTimeRating(medianHours);

  return {
    averageHours,
    medianHours,
    p90Hours,
    rating,
  };
}

function getLeadTimeRating(medianHours: number): DoraRating {
  // Elite: Less than one hour
  if (medianHours < 1) return 'elite';
  // High: Less than one day
  if (medianHours < 24) return 'high';
  // Medium: Less than one week
  if (medianHours < 168) return 'medium';
  // Low: More than one week
  return 'low';
}

function calculateChangeFailureRate(
  deployments: GitHubDeployment[],
  incidents: LinearIssue[]
): ChangeFailureRate {
  const successfulDeployments = deployments.filter(d => d.status === 'success');
  const failedDeployments = deployments.filter(d => d.status === 'failure');
  const totalDeployments = successfulDeployments.length + failedDeployments.length;

  // Count incidents that occurred after deployments as change failures
  const deploymentFailures = failedDeployments.length + incidents.length;

  const percentage = totalDeployments > 0
    ? (deploymentFailures / totalDeployments) * 100
    : 0;

  const rating = getChangeFailureRateRating(percentage);

  return {
    percentage,
    failedDeployments: deploymentFailures,
    totalDeployments,
    rating,
  };
}

function getChangeFailureRateRating(percentage: number): DoraRating {
  // Elite: 0-5%
  if (percentage <= 5) return 'elite';
  // High: 5-10%
  if (percentage <= 10) return 'high';
  // Medium: 10-15%
  if (percentage <= 15) return 'medium';
  // Low: More than 15%
  return 'low';
}

function calculateMTTR(
  incidents: LinearIssue[],
  postmortems: SlabPostmortem[]
): MTTR {
  const resolutionTimes: number[] = [];

  // Use incident cycle times from Linear
  for (const incident of incidents) {
    if (incident.cycleTime !== undefined) {
      resolutionTimes.push(incident.cycleTime);
    }
  }

  // Also use postmortem resolution times from Slab
  for (const postmortem of postmortems) {
    resolutionTimes.push(postmortem.timeToResolve);
  }

  if (resolutionTimes.length === 0) {
    return {
      averageHours: 0,
      medianHours: 0,
      incidents: 0,
      rating: 'elite', // No incidents is elite
    };
  }

  const sortedTimes = [...resolutionTimes].sort((a, b) => a - b);
  const averageHours = resolutionTimes.reduce((a, b) => a + b, 0) / resolutionTimes.length;
  const medianHours = sortedTimes[Math.floor(sortedTimes.length / 2)];

  const rating = getMTTRRating(medianHours);

  return {
    averageHours,
    medianHours,
    incidents: incidents.length + postmortems.length,
    rating,
  };
}

function getMTTRRating(medianHours: number): DoraRating {
  // Elite: Less than one hour
  if (medianHours < 1) return 'elite';
  // High: Less than one day
  if (medianHours < 24) return 'high';
  // Medium: Less than one week
  if (medianHours < 168) return 'medium';
  // Low: More than one week
  return 'low';
}

export function getOverallRating(metrics: DoraMetrics): DoraRating {
  const ratings = [
    metrics.deploymentFrequency.rating,
    metrics.leadTime.rating,
    metrics.changeFailureRate.rating,
    metrics.mttr.rating,
  ];

  const ratingValues: Record<DoraRating, number> = {
    elite: 4,
    high: 3,
    medium: 2,
    low: 1,
  };

  const averageValue = ratings.reduce((sum, r) => sum + ratingValues[r], 0) / ratings.length;

  if (averageValue >= 3.5) return 'elite';
  if (averageValue >= 2.5) return 'high';
  if (averageValue >= 1.5) return 'medium';
  return 'low';
}

export function formatRating(rating: DoraRating): string {
  const labels: Record<DoraRating, string> = {
    elite: 'Elite Performer',
    high: 'High Performer',
    medium: 'Medium Performer',
    low: 'Low Performer',
  };
  return labels[rating];
}

File: metrics_dashboard/dora.py
"""DORA metrics calculation."""

from datetime import datetime
from statistics import median

from metrics_dashboard.models import (
    ChangeFailureRate,
    DataFetchResult,
    DeploymentFrequency,
    DoraMetrics,
    DoraRating,
    GitHubDeployment,
    GitHubPullRequest,
    Incident,
    LeadTime,
    MetricsPeriod,
    MTTR,
)


def calculate_dora_metrics(data: DataFetchResult, period: MetricsPeriod) -> DoraMetrics:
    """Calculate all DORA metrics from fetched data."""
    # Filter to only change-related incidents for DORA metrics
    change_incidents = [inc for inc in data.incidents if inc.is_change_related]

    return DoraMetrics(
        deployment_frequency=calculate_deployment_frequency(data.deployments, period),
        lead_time=calculate_lead_time(data.pull_requests),
        change_failure_rate=calculate_change_failure_rate(data.deployments, change_incidents),
        mttr=calculate_mttr(change_incidents),
        period=period,
        generated_at=datetime.now(),
    )


def calculate_deployment_frequency(
    deployments: list[GitHubDeployment], period: MetricsPeriod
) -> DeploymentFrequency:
    """Calculate deployment frequency metric."""
    successful = [d for d in deployments if d.status == "success"]
    total = len(successful)

    days = max((period.end_date - period.start_date).days, 1)
    per_day = total / days
    per_week = per_day * 7

    rating = _get_deployment_frequency_rating(per_day)

    return DeploymentFrequency(
        deployments_per_day=per_day,
        deployments_per_week=per_week,
        total_deployments=total,
        rating=rating,
    )


def _get_deployment_frequency_rating(per_day: float) -> DoraRating:
    """Get rating for deployment frequency."""
    if per_day >= 1:
        return DoraRating.ELITE
    elif per_day >= 1 / 7:
        return DoraRating.HIGH
    elif per_day >= 1 / 30:
        return DoraRating.MEDIUM
    return DoraRating.LOW


def calculate_lead_time(pull_requests: list[GitHubPullRequest]) -> LeadTime:
    """Calculate lead time for changes metric."""
    lead_times: list[float] = []

    for pr in pull_requests:
        if not pr.merged_at:
            continue

        start_time = pr.first_commit_at or pr.created_at
        hours = (pr.merged_at - start_time).total_seconds() / 3600
        lead_times.append(hours)

    if not lead_times:
        return LeadTime(
            average_hours=0,
            median_hours=0,
            p90_hours=0,
            rating=DoraRating.LOW,
        )

    sorted_times = sorted(lead_times)
    avg = sum(lead_times) / len(lead_times)
    med = median(lead_times)
    p90_idx = int(len(sorted_times) * 0.9)
    p90 = sorted_times[p90_idx] if p90_idx < len(sorted_times) else med

    rating = _get_lead_time_rating(med)

    return LeadTime(
        average_hours=avg,
        median_hours=med,
        p90_hours=p90,
        rating=rating,
    )


def _get_lead_time_rating(median_hours: float) -> DoraRating:
    """Get rating for lead time."""
    if median_hours < 1:
        return DoraRating.ELITE
    elif median_hours < 24:
        return DoraRating.HIGH
    elif median_hours < 168:  # 1 week
        return DoraRating.MEDIUM
    return DoraRating.LOW


def calculate_change_failure_rate(
    deployments: list[GitHubDeployment], incidents: list[Incident]
) -> ChangeFailureRate:
    """Calculate change failure rate metric.

    Per DORA definition: The percentage of deployments causing a failure
    in production that requires remediation (rollback, hotfix, patch).

    We count incidents from incident.io that are marked as change-related.
    """
    successful_deployments = len([d for d in deployments if d.status == "success"])
    total_deployments = successful_deployments

    if total_deployments == 0:
        return ChangeFailureRate(
            percentage=0,
            failed_changes=0,
            total_deployments=0,
            rating=DoraRating.ELITE,
        )

    # Count change-related incidents as failed changes
    failed_changes = len(incidents)

    percentage = (failed_changes / total_deployments * 100) if total_deployments > 0 else 0

    rating = _get_change_failure_rate_rating(percentage)

    return ChangeFailureRate(
        percentage=percentage,
        failed_changes=failed_changes,
        total_deployments=total_deployments,
        rating=rating,
    )


def _get_change_failure_rate_rating(percentage: float) -> DoraRating:
    """Get rating for change failure rate."""
    if percentage <= 5:
        return DoraRating.ELITE
    elif percentage <= 10:
        return DoraRating.HIGH
    elif percentage <= 15:
        return DoraRating.MEDIUM
    return DoraRating.LOW


def calculate_mttr(incidents: list[Incident]) -> MTTR:
    """Calculate mean time to recovery metric.

    Per DORA definition: How long it takes to restore service when a
    service incident or defect that impacts users occurs.
    """
    resolution_times: list[float] = []

    for incident in incidents:
        if incident.time_to_resolve_hours is not None:
            resolution_times.append(incident.time_to_resolve_hours)

    if not resolution_times:
        return MTTR(
            average_hours=0,
            median_hours=0,
            incidents=0,
            rating=DoraRating.ELITE,  # No incidents is elite
        )

    avg = sum(resolution_times) / len(resolution_times)
    med = median(resolution_times)
    rating = _get_mttr_rating(med)

    return MTTR(
        average_hours=avg,
        median_hours=med,
        incidents=len(incidents),
        rating=rating,
    )


def _get_mttr_rating(median_hours: float) -> DoraRating:
    """Get rating for MTTR."""
    if median_hours < 1:
        return DoraRating.ELITE
    elif median_hours < 24:
        return DoraRating.HIGH
    elif median_hours < 168:  # 1 week
        return DoraRating.MEDIUM
    return DoraRating.LOW


def get_overall_rating(metrics: DoraMetrics) -> DoraRating:
    """Calculate overall DORA rating."""
    ratings = [
        metrics.deployment_frequency.rating,
        metrics.lead_time.rating,
        metrics.change_failure_rate.rating,
        metrics.mttr.rating,
    ]

    values = {
        DoraRating.ELITE: 4,
        DoraRating.HIGH: 3,
        DoraRating.MEDIUM: 2,
        DoraRating.LOW: 1,
    }

    avg = sum(values[r] for r in ratings) / len(ratings)

    if avg >= 3.5:
        return DoraRating.ELITE
    elif avg >= 2.5:
        return DoraRating.HIGH
    elif avg >= 1.5:
        return DoraRating.MEDIUM
    return DoraRating.LOW


def format_rating(rating: DoraRating) -> str:
    """Format rating for display."""
    labels = {
        DoraRating.ELITE: "Elite Performer",
        DoraRating.HIGH: "High Performer",
        DoraRating.MEDIUM: "Medium Performer",
        DoraRating.LOW: "Low Performer",
    }
    return labels[rating]
